scipy/optimize/_optimize.py

```diff
@@ -131,6 +132,59 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
     return s
 
 
+def _wrap_callback(callback, method=None):
+    """Wrap a user-provided callback so that attributes can be attached."""
+    if callback is None or method not in {'nelder-mead', 'trust-constr', None}:
+        return callback  # don't wrap
+
+    p1 = (inspect.Parameter('intermediate_result',
+                            kind=inspect.Parameter.POSITIONAL_OR_KEYWORD),)
+    p2 = (inspect.Parameter('intermediate_result',
+                            kind=inspect.Parameter.KEYWORD_ONLY),)
```

Comment:
I can see some discussion about keyword only. That would be my preference but this works to as in the end we only call the callback using the parameter's name.

Reply:
Yes, I had it keyword-only originally, but Andrew asked for it to positional or keyword.